### PR TITLE
fix: prevent time dims from being wiped when reopening modal (TECH-1061)

### PR DESCRIPTION
**Pull request by martinkrulltott**
Implements [TECH-1061](https://dhis2.atlassian.net/browse/TECH-1061)

Prevents time dimensions from being wiped when the modal is reopened. Incomplete start/end dates are still removed as they should.

Needs to be tested thoroughly.

### CODE REVIEW

File: src/components/Dialogs/PeriodDimension/PeriodDimension.js [hunk 1]
```diff
@@ -101,7 +101,11 @@ export const PeriodDimension = ({ dimension, onClose }) => {
     }, [presets])
 
     useEffect(() => {
-        updatePeriodDimensionItems(startEndDate ? [{ id: startEndDate }] : [])
+        if (startEndDate) {
+            updatePeriodDimensionItems([{ id: startEndDate }])
+        } else if (!presets?.length) {
+            updatePeriodDimensionItems([])
+        }
     }, [startEndDate])
```
[the remark this answers is not in the document]
Comment (jenniferarnesen) [reply src/components/Dialogs/PeriodDimension/PeriodDimension.js:109]: Actually, maybe this effect and the one above it should be combined?

[the remark this answers is not in the document]
Comment (martinkrulltott) [reply src/components/Dialogs/PeriodDimension/PeriodDimension.js:109]: I tried the following two 
```
    useEffect(() => {
        if (startEndDate) {
            updatePeriodDimensionItems([{ id: startEndDate }])
        } else if (!presets?.length) {
            updatePeriodDimensionItems([])
        } else if (presets) {
            updatePeriodDimensionItems(presets)
        }
    }, [startEndDate, presets])
```

```
    useEffect(() => {
        if (presets) {
            updatePeriodDimensionItems(presets)
        } else if (startEndDate) {
            updatePeriodDimensionItems([{ id: startEndDate }])
        } else if (!presets?.length) {
            updatePeriodDimensionItems([])
        }
    }, [presets, startEndDate])
```

But neither works. In both cases, the preset will overwrite the startEndDate, even when the startEndDate is added last. E.g. go to the period dimension, select "last 12 months", change to defined dates and enter two dates, click update - Last 12 months is still in effect.

[the remark this answers is not in the document]
Comment (jenniferarnesen) [reply src/components/Dialogs/PeriodDimension/PeriodDimension.js:109]: would this work? 
```
    useEffect(() => {
        if (entryMethod === OPTION_START_END_DATES) {
            updatePeriodDimensionItems(
                startEndDate ? [{ id: startEndDate }] : []
            )
        } else {
            updatePeriodDimensionItems(presets)
        }
    }, [entryMethod, presets, startEndDate])
```

[the remark this answers is not in the document]
Comment (martinkrulltott) [reply src/components/Dialogs/PeriodDimension/PeriodDimension.js]: refactored, useEffects removed 🙂

File: src/components/Dialogs/PeriodDimension/StartEndDate.js [hunk 2]
```diff
@@ -10,7 +10,11 @@ export const StartEndDate = ({ value, setValue }) => {
     const [endDate, setEndDate] = useState(endDateStr)
 
     useEffect(() => {
-        setValue(startDate && endDate ? `${startDate}_${endDate}` : '')
+        if (startDate && endDate) {
```
Comment (jenniferarnesen) [inline src/components/Dialogs/PeriodDimension/StartEndDate.js:13]: What about this solution, that doesn't use `value` in useEffect (which will raise the exhaustive dependencies warning):

```
    useEffect(() => {
        if (startDate && endDate) {
            setValue(`${startDate}_${endDate}`)
        } else if (!startDate || !endDate) {
            setValue(null)
        }
    }, [startDate, endDate])
```

and correspondingly in PeriodDimension:
```
setValue={(value) => {
    if (value === null && selectedIds.length) {
        updatePeriodDimensionItems([])
    } else if (value && value !== selectedIds[0]) {
        updatePeriodDimensionItems([{ id: value }])
    }
}}
```

This way only PeriodDimension maintains the management control of selectedIds.

**Review by jenniferarnesen — APPROVED** [review]

File: src/components/Dialogs/PeriodDimension/StartEndDate.js [hunk 3]
```diff
@@ -10,7 +10,7 @@ export const StartEndDate = ({ value, setValue }) => {
     const [endDate, setEndDate] = useState(endDateStr)
 
     useEffect(() => {
-        setValue(startDate && endDate ? `${startDate}_${endDate}` : '')
+        setValue(startDate && endDate ? `${startDate}_${endDate}` : null)
```
Comment (jenniferarnesen) [inline src/components/Dialogs/PeriodDimension/StartEndDate.js:13]: If you want, you can stick with empty string for both here and line 152 of PeriodDImension, or null in both places?

Comment (martinkrulltott) [inline src/components/Dialogs/PeriodDimension/StartEndDate.js:13]:
```suggestion
        setValue(startDate && endDate ? `${startDate}_${endDate}` : '')
```

**Pull request merged**